gdk4/src/toplevel.rs

```diff
@@ -1,18 +1,51 @@
 // Take a look at the license at the top of the repository in the LICENSE file.
 
-use crate::{Event, Toplevel};
+use crate::{Event, Toplevel, ToplevelSize};
+use glib::signal::connect_raw;
+use glib::signal::SignalHandlerId;
 use glib::translate::*;
 use glib::IsA;
+use std::boxed::Box as Box_;
+use std::mem::transmute;
 
 pub trait ToplevelExtManual {
     #[doc(alias = "gdk_toplevel_inhibit_system_shortcuts")]
     fn inhibit_system_shortcuts<P: AsRef<Event>>(&self, event: Option<&P>);
 
     #[doc(alias = "gdk_toplevel_show_window_menu")]
     fn show_window_menu<P: AsRef<Event>>(&self, event: &P) -> bool;
+    fn connect_compute_size<F: Fn(&Toplevel, &ToplevelSize) + 'static>(
+        &self,
+        f: F,
+    ) -> SignalHandlerId;
 }
 
 impl<O: IsA<Toplevel>> ToplevelExtManual for O {
+    fn connect_compute_size<F: Fn(&Toplevel, &ToplevelSize) + 'static>(
+        &self,
+        f: F,
+    ) -> SignalHandlerId {
+        unsafe extern "C" fn compute_size_trampoline<F: Fn(&Toplevel, &ToplevelSize) + 'static>(
+            this: *mut ffi::GdkToplevel,
+            size: *mut ffi::GdkToplevelSize,
```

Comment:
Not sure how If that's the right to handle size here as it's annotated as `out caller-allocates` does that means the signal should be returning a ToplevelSize ?

Reply:
I don't know what that means in the context of signals.

But is it intended that the signal handler fills in the `size`?

Reply:
Yes, as the C code uses it like this
```c
void
gdk_toplevel_notify_compute_size (GdkToplevel     *toplevel,
                                  GdkToplevelSize *size)
{
  g_signal_emit (toplevel, signals[COMPUTE_SIZE], 0, size);
  gdk_toplevel_size_validate (size);
}
```

Reply:
So you need getters and setters for the values in `ToplevelSize`, and make sure that the signal handlers gets a mutable reference.

The `out caller-allocates` seems strange here but I guess makes sense: the value is allocated by the caller to `g_signal_emit()` and it expects the signal to fill in things.

Can you check if all callers initialize the `ToplevelSize` with something meaningful (like zeroes) or if they just contain random uninitialized memory?

Reply:
Can you check if all callers initialize the ToplevelSize with something meaningful (like zeroes) or if they just contain random uninitialized memory?
> it's an initialized ToplevelSize with `gdk_toplevel_size_init(&size, bound_width, bound_height);`

> So you need getters and setters for the values in ToplevelSize, and make sure that the signal handlers gets a mutable reference.
I will have a look at the various gdk backends and see which "getters"/"setters" they use and add them

Reply:
Ok, let me know. I'll do another review after that is solved